crates/language/src/language_registry.rs

```diff
@@ -709,21 +706,44 @@ impl LanguageRegistry {
             })
         });
         self.find_matching_language(move |language_name, config, current_best_match| {
+            let filename = filename?;
+
+            let matching_suffix_len = |acc: usize, suffix: &String| {
+                let ext = ".".to_string() + suffix;
+                if filename.ends_with(&ext) || filename == suffix {
+                    acc.max(suffix.len())
+                } else {
+                    acc
+                }
+            };
+
             let path_matches_default_suffix = || {
-                config
-                    .path_suffixes
-                    .iter()
-                    .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())))
+                let len = config.path_suffixes.iter().fold(0, matching_suffix_len);
+                Some(len).filter(|len| len > &0)
             };
+
             let path_matches_custom_suffix = || {
                 user_file_types
                     .and_then(|types| types.get(language_name.as_ref()))
-                    .map_or(false, |custom_suffixes| {
-                        path_suffixes_candidates
-                            .iter()
-                            .any(|suffix| custom_suffixes.is_match_candidate(suffix))
+                    .map_or(None, |custom_suffixes| {
+                        let matches = custom_suffixes.glob.matches(filename);
+
+                        let len = if !matches.is_empty() {
+                            // if the glob matched, use the length of its pattern
+                            matches
+                                .iter()
+                                .filter_map(|i| custom_suffixes.patterns.get(*i))
+                                .fold(0, |acc, suffix| acc.max(suffix.len()))
+                        } else {
+                            // otherwise try to use the glob patterns as extensions, and
+                            // see if the file ends with any of them
+                            custom_suffixes.patterns.iter().fold(0, matching_suffix_len)
+                        };
```

Comment:
This is reason I added the `LanguageCustomFileTypes` struct: the glob matcher could be made up of more than 1 glob pattern, and `is_match()` and `is_match_candidate()` only say if any pattern matched. `matches()` returns which patterns matched, which then allows us to examine each matched pattern to see how long it was.

Reply:
I don't think you want to check the length of the glob pattern; we want the length of the suffix that matched (either the length of the file extension, the length of the filename, or the length of the entire path). So you don't need to store off the glob patterns.

Reply:
Done!